### PR TITLE
Handle upp identifiers

### PR DESCRIPTION
**Pull request by duffj**
[UPP-1317](https://jira.ft.com/browse/UPP-1317)

FYI the tests won't pass unless you switch to the correct branch of the test dependencies for other FT services. This script will help:

```
#!/usr/bin/env bash
echo memberships-rw-neo4j; cd $HOME/dev/go/src/github.com/Financial-Times/memberships-rw-neo4j && git checkout feature/alternativeIdentifiers && git pull origin && go clean
echo organisations-rw-neo4j; cd $HOME/dev/go/src/github.com/Financial-Times/organisations-rw-neo4j && git checkout update-identifier-structure && git pull origin && go clean
echo people-rw-neo4j; cd $HOME/dev/go/src/github.com/Financial-Times/people-rw-neo4j && git checkout add-identifiers && git pull origin && go clean
echo roles-rw-neo4j; cd $HOME/dev/go/src/github.com/Financial-Times/roles-rw-neo4j && git checkout newIdentifierFormat && git pull origin && go clean
```

### CODE REVIEW

**Comment by duffj** [comment]
CircleCI checks will fail until the following PR is merged: https://github.com/Financial-Times/people-rw-neo4j/pull/17

File: people/cypher.go [hunk 1]
```diff
@@ -117,6 +119,7 @@ func (pcw CypherDriver) Read(uuid string) (person Person, found bool, err error)
 	}
 	log.Infof("CypherResult ReadPeople for uuid: %s was: %+v", uuid, results)
 	if (len(results)) == 0 || len(results[0].Rs) == 0 {
+		log.WithFields(log.Fields{"uuid":uuid.String()}).Debug("Result not found")
```
Comment (BerniVarga) [inline people/cypher.go]: Is this still needed?

File: people/handlers.go [hunk 2]
```diff
@@ -64,15 +68,16 @@ func MethodNotAllowedHandler(w http.ResponseWriter, r *http.Request) {
...
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
 		return
 	}
-	person, found, err := PeopleDriver.Read(uuid)
+
+	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
+	person, found, err := PeopleDriver.Read(requestedId)
 	if err != nil {
+		log.WithFields(log.Fields{"requestedId": requestedId, "err": err}).Debug("Redirecting...")
```
Comment (BerniVarga) [inline people/handlers.go]: And this?

Comment (BerniVarga) [inline people/handlers.go]: And this?

File: people/handlers.go [hunk 3]
```diff
@@ -83,6 +88,19 @@ func GetPerson(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	// if the client requested the non-canonical UUID, then we redirect them to the URL for the canonical ID.
+	canonicalId, err := extractCanonicalUUID(person)
+	if err != nil {
+		log.WithFields(log.Fields{"ID": person.ID, "err": err}).Error("Error reading canonical ID")
+	}
+	if !uuid.Equal(canonicalId, requestedId) {
+		log.WithFields(log.Fields{"canonicalId": canonicalId, "requestedId": requestedId}).Debug("Redirecting...")
```
Comment (BerniVarga) [inline people/handlers.go]: Do we really have to log when a redirect url will be returned? The status code and location header will indicate this.

File: people/handlers.go [hunk 4]
```diff
@@ -91,3 +109,15 @@ func GetPerson(w http.ResponseWriter, r *http.Request) {
 		w.Write([]byte(`{"message":"Person could not be retrieved, err=` + err.Error() + `"}`))
 	}
 }
+
+// extract the UUID from the person ID URL by taking the last element of the path.
+func extractCanonicalUUID(person Person) (uuid.UUID, error) {
+	u, err := url.Parse(person.ID)
+	log.WithFields(log.Fields{"u": u}).Debug("Parsed URL")
```
Comment (BerniVarga) [inline people/handlers.go]: Debug msg

File: people/handlers.go (hunk 3 above)
Comment (nwrigh) [inline people/handlers.go]: I think it would be easier for people to see in the logs at least at the start of using this

Comment (BerniVarga) [inline people/handlers.go]: Isn't the status code for a request logged anyway? But i don't mind the msg, it can remain. :)

**Comment by BerniVarga** [comment]
+1

**Pull request merged**